happi/cli.py

```diff
@@ -1075,6 +1077,73 @@ def audit(
         click.echo(f'# devices failed: {len(unique_fails)} / {len(results)}')
 
 
+@happi_cli.command()
+@click.pass_context
+@click.option('--fix-optional/--ignore-optinal', 'fix_optional', default=False,
+              help='Also prompt for user input on optional information')
+@click.option('--glob/--regex', 'use_glob', default=True,
+              help='Use glob (default) or regex style search terms. '
+              'Only relevant if search_criteria are provided.')
+@click.argument('search_criteria', nargs=-1)
+def repair(
+    ctx,
+    fix_optional: bool,
+    use_glob: bool,
+    search_criteria: tuple[str, ...]
+):
+    """
+    Repair the database.
+
+    Repairs all entries matching SEARCH_CRITERIA, repairs entire database otherwise.
+    """
+    logger.debug('starting repair block')
+
+    client: happi.Client = ctx.obj
+    if search_criteria:
+        results = search_parser(client, use_glob, search_criteria)
+    else:
+        # run repair on all items
+        results = search_parser(client, True, '*')
+
+    for res in results:
+        # fix mandatory info with missing defaults
+        req_info = (info for info in find_unfilled_mandatory_info(res))
+
+        if fix_optional:
+            empty_opt = find_unfilled_optional_info(res)
+            req_info = itertools.chain(req_info, empty_opt)
+
+        try:
+            req_field = next(req_info)
+        except StopIteration:
+            req_field = False
+
+        res_name = res['_id']
+
+        # fix each mandatory field
+        logger.info(f'repairing ({res_name})...')
+        while req_field:
+            info = res.item._info_attrs[req_field]
+            req_value = prompt_for_entry(info)
+            info.enforce_value(req_value)
+            setattr(res.item, req_field, req_value)
+            try:
+                req_field = next(req_info)
+            except StopIteration:
+                break
+
+        # re-save after creating container
+        try:
+            # will save optional data with defaults
+            # optional data without defaults saved as null
+            res.item.save()
+        except KeyboardInterrupt:
+            # Finish the current save if interrupted
+            logger.warning('caught keboard interrupt, finishing')
```

Comment:
```suggestion
            logger.warning('caught keyboard interrupt, finishing')
```

Reply:
Is the expectation that an interrupted `save()` could be repaired by trying again after `KeyboardInterrupt`?

Reply:
My experience showed that an interrupted `save()` may end up with an empty `db.json`, and saving again would write the file properly.  I'm not entirely sure that's the intention, but that worked for me.

Reply:
Oh, interesting... That makes me think we should be a lot more careful with `save()` in general.

It might be good to suppress `KeyboardInterrupt` during `save()`, or have the backend do a safer "write-and-move" operation... Let's chat about this maybe tomorrow